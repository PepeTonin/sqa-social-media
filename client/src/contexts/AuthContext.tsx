"use client";

import { createContext, useContext, useState, useEffect, ReactNode } from "react";
import { getUser, saveUser, removeUser } from "@/lib/localStorage";
import { User } from "@/service/types";


interface AuthContextType {
  user: User | null;
  isAuthenticated: boolean;
  isLoading: boolean;
  login: (user: User) => void;
  logout: () => void;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthProvider({ children }: { children: ReactNode }) {
  const [user, setUser] = useState<User | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const storedUser = getUser();
    if (storedUser) {
      setUser(storedUser);
    }
    setIsLoading(false);
  }, []);

  function login(userData: User) {
    setUser(userData);
    saveUser(userData);
  }

  function logout() {
    setUser(null);
    removeUser();
  }

  const value: AuthContextType = {
    user,
    isAuthenticated: user !== null,
    isLoading,
    login,
    logout,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export function useAuth() {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
}

