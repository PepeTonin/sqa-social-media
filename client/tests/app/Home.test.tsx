import { AuthProvider } from "../../src/contexts/AuthContext";
import { render, screen } from "@testing-library/react";
import Home from "@/app/page";
import api from "../../src/service/api";

jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: jest.fn(),
    replace: jest.fn(),
    prefetch: jest.fn(),
    back: jest.fn(),
    forward: jest.fn(),
    refresh: jest.fn(),
    pathname: "/",
  }),
  usePathname: () => "/",
}));

jest.mock("../../src/service/api");

const mockedApi = api as jest.Mocked<typeof api>;

describe("App - Home", () => {
  test("deve renderizar o nÃºmero de posts corretamente de acordo com a quantidade de posts retornada pela API", async () => {
    const mockResponse = {
      data: {
        posts: [
          { id: 1, title: "Post 1", body: "Content 1", liked: false },
          { id: 2, title: "Post 2", body: "Content 2", liked: false },
        ],
      },
    };

    mockedApi.get.mockResolvedValueOnce(mockResponse);

    const { asFragment } = render(
      <AuthProvider>
        <Home />
      </AuthProvider>
    );

    await screen.findAllByRole("listitem");
    expect(screen.getAllByRole("listitem").length).toBe(2);
    expect(asFragment()).toMatchSnapshot();
  });
});
