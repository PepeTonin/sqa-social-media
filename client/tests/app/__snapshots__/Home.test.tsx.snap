// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`App - Home deve renderizar o n√∫mero de posts corretamente de acordo com a quantidade de posts retornada pela API 1`] = `
<DocumentFragment>
  <div
    style="min-height: 100vh; background: var(--background);"
  >
    <header
      style="background: var(--card-bg); border-bottom: 1px solid var(--border); padding: 1rem 2rem;"
    >
      <div
        style="max-width: 1200px; margin: 0px auto; display: flex; justify-content: space-between; align-items: center;"
      >
        <h1
          style="font-size: 1.5rem; font-weight: bold; cursor: pointer; color: var(--primary);"
        >
          SQA Social Media
        </h1>
        <div
          style="display: flex; gap: 1rem; align-items: center;"
        >
          <button
            style="background: transparent; color: var(--foreground); border: 1px solid var(--border); padding: 0.5rem 1rem; border-radius: 0.375rem; cursor: pointer; font-weight: 500; font-size: 1rem; transition: all 0.2s; opacity: 1;"
          >
            Entrar
          </button>
          <button
            style="background: var(--primary); color: white; padding: 0.5rem 1rem; border-radius: 0.375rem; cursor: pointer; font-weight: 500; font-size: 1rem; transition: all 0.2s; opacity: 1;"
          >
            Criar Conta
          </button>
        </div>
      </div>
    </header>
    <main
      style="max-width: 800px; margin: 0px auto; padding: 2rem;"
    >
      <h1
        style="font-size: 2rem; font-weight: bold; margin-bottom: 2rem; color: var(--foreground);"
      >
        Feed de Posts
      </h1>
      <div>
        <div
          role="listitem"
          style="background: var(--card-bg); border: 1px solid var(--border); border-radius: 0.5rem; padding: 1.5rem; margin-bottom: 1rem; transition: all 0.2s;"
        >
          <h2
            style="font-size: 1.25rem; font-weight: 600; margin-bottom: 0.75rem; color: var(--foreground);"
          >
            Post 1
          </h2>
          <p
            style="color: var(--foreground); opacity: 0.9; line-height: 1.6; margin-bottom: 1rem;"
          >
            Content 1
          </p>
          <div
            style="display: flex; justify-content: flex-end; align-items: center; margin-top: 1rem;"
          >
            <button
              style="background: transparent; color: var(--foreground); border: 2px solid var(--border); padding: 0.5rem 1.25rem; border-radius: 0.375rem; cursor: pointer; font-weight: 500; display: flex; align-items: center; gap: 0.5rem; transition: all 0.2s; opacity: 1;"
            >
              <span
                style="font-size: 1.25rem;"
              >
                ü§ç
              </span>
              <span>
                Curtir
              </span>
            </button>
          </div>
        </div>
        <div
          role="listitem"
          style="background: var(--card-bg); border: 1px solid var(--border); border-radius: 0.5rem; padding: 1.5rem; margin-bottom: 1rem; transition: all 0.2s;"
        >
          <h2
            style="font-size: 1.25rem; font-weight: 600; margin-bottom: 0.75rem; color: var(--foreground);"
          >
            Post 2
          </h2>
          <p
            style="color: var(--foreground); opacity: 0.9; line-height: 1.6; margin-bottom: 1rem;"
          >
            Content 2
          </p>
          <div
            style="display: flex; justify-content: flex-end; align-items: center; margin-top: 1rem;"
          >
            <button
              style="background: transparent; color: var(--foreground); border: 2px solid var(--border); padding: 0.5rem 1.25rem; border-radius: 0.375rem; cursor: pointer; font-weight: 500; display: flex; align-items: center; gap: 0.5rem; transition: all 0.2s; opacity: 1;"
            >
              <span
                style="font-size: 1.25rem;"
              >
                ü§ç
              </span>
              <span>
                Curtir
              </span>
            </button>
          </div>
        </div>
      </div>
      <p
        style="text-align: center; margin-top: 2rem; color: var(--foreground); opacity: 0.7;"
      >
        Voc√™ chegou ao fim dos posts!
      </p>
    </main>
  </div>
</DocumentFragment>
`;
